### ISSUE TITLE
sql: TypeCompare should return false for DValArg

### ISSUE THREAD

**Issue by mjibson** [cockroachdb/cockroach#4088]
See: https://github.com/cockroachdb/cockroach/pull/4082#issuecomment-178742589

**Comment by mjibson** [comment]
The problem is when this returns false or panics, the following prepared statement fails:

```
INSERT INTO d.t VALUES ($1), (1)
```

because values.go makes sure that all columns for inserts with multiple rows share the same type. Here the `1` (an int) and the `$1` (a val arg) are different types. The MapArgs variable isn't populated at this point, so it's not possible to infer the type for `$1` yet.

Making TypeCompare always return `true` for DValArg won't work either, because you could have a situation like:

```
INSERT INTO d.t VALUES ($1), (1), ("s")
```

Since values.go uses the first row as the type, it would compare the valarg to both `1` and then `"s"` and return true for both, even though they really do have differing types.

One good solution would be to have MapArgs populated by this point so that TypeComparison can use the val arg's eventual type and do a correct type compare. This would require refactoring the prepare implementation of INSERT to populate MapArgs before it calls into Values.

**Comment by andreimatei** [comment]
Indeed, I think the types of placeholders should be inferred by the time they're needed by type checking. A few of us want to refactor how inference works, and make it a separate phase of semantic analysis. We're gonna send a proposal today.

**Comment by knz** [comment]
Fixed by #4121.
